contrib/jms/src/test/java/com/google/cloud/dataflow/contrib/jms/JmsIOTest.java

```diff
@@ -0,0 +1,140 @@
+        broker.setBrokerName("localhost");
+        broker.start();
+
+        // create JMS connection factory
+        System.out.println("Create JMS connection factory on " + BROKER_URL);
+        connectionFactory = new ActiveMQConnectionFactory(BROKER_URL);
+    }
+
+    @Test
+    public void testReadSingleMessage() throws Exception {
+
+        // produce message
+        System.out.println("Producing a test message on the broker ...");
+        Connection connection = connectionFactory.createConnection();
+        Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+        MessageProducer producer = session.createProducer(session.createQueue("test"));
+        TextMessage message = session.createTextMessage("This Is A Test");
+        producer.send(message);
+        producer.send(message);
+        producer.send(message);
+        producer.send(message);
+        producer.send(message);
+        producer.send(message);
+        producer.close();
+        session.close();
+        connection.close();
+
+        System.out.println("Creating test pipeline");
+        Pipeline pipeline = TestPipeline.create();
+
+        // read from the queue
+        System.out.println("Reading from the JMS queue and create PCollection");
+        PCollection<String> output =  pipeline.apply(
+                JmsIO.Read
+                        .connectionFactory(connectionFactory)
+                        .queue("test")
+                        .maxNumMessages(5));
+
+        System.out.println("Starting the pipeline");
+        pipeline.run();
+
+        System.out.println("Check assertion");
+        DataflowAssert.that(output).containsInAnyOrder(new String[]{ "test "});
+
+        System.out.println(output);
+    }
+
+    @Test
+    public void testWriteMessage() throws Exception {
+        System.out.println("Create test pipeline");
+        Pipeline pipeline = TestPipeline.create();
+
+        System.out.println("Create PCollection");
+        ArrayList<String> data = new ArrayList<>();
+        data.add("Test");
+        PCollection<String> input = pipeline.apply(Create.of(data).withCoder(StringUtf8Coder.of()));
+
+        System.out.println("Write PCollection to the JMS queue");
+        JmsIO.Write.Bound<String> write = JmsIO.Write
+                .connectionFactory(connectionFactory)
+                .queue("test");
+
+        input.apply(write);
+
+        pipeline.run();
+
+        Connection connection = connectionFactory.createConnection();
+        connection.start();
+        Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+        MessageConsumer messageConsumer = session.createConsumer(session.createQueue("test"));
+        TextMessage message = (TextMessage) messageConsumer.receive(10000);
+        Assert.assertEquals(message.getText(), "Test");
+    }
+
+    @After
+    public void stopBroker() throws Exception {
+        broker.stop();
+    }
+
```

Comment:
I would like to see a few more unit tests in this file. Things like building the Read and Write transform works, and that validation catches invalid Read and Write transforms. For examples, maybe see https://github.com/GoogleCloudPlatform/DataflowJavaSDK/blob/master/sdk/src/test/java/com/google/cloud/dataflow/sdk/io/bigtable/BigtableIOTest.java#L159 and related tests.

Basically, improving test coverage of the relevant lines in the file. In general, the more coverage the better tho obviously we should be aware of diminishing returns.